### PR TITLE
Minor: Improve HashJoinStream docstrings

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?
Related to https://github.com/apache/arrow-datafusion/pull/8020

## Rationale for this change

While reviewing https://github.com/apache/arrow-datafusion/pull/8020 I studied the join code and want to encode my learning as doc comments for both by future self and hopefully it will be useful to others as well

## What changes are included in this PR?

1. Update doc comments in `HashJoinStream`

## Are these changes tested?

By doc CI test
## Are there any user-facing changes?
Better docstrings

### CODE REVIEW

File: datafusion/physical-plan/src/joins/hash_join.rs [hunk 1]
```diff
@@ -1003,13 +1028,14 @@ impl HashJoinStream {
         cx: &mut std::task::Context<'_>,
     ) -> Poll<Option<Result<RecordBatch>>> {
         let build_timer = self.join_metrics.build_time.timer();
+        // build hash table from left (build) side, if not yet done
```
Comment (comphead) [inline datafusion/physical-plan/src/joins/hash_join.rs:1031]: that looks great @alamb 
Thanks for leaving comments where is build and where is a probe.

File: datafusion/physical-plan/src/joins/hash_join.rs [hunk 2]
```diff
@@ -747,27 +747,38 @@ where
     Ok(())
 }
 
-/// A stream that issues [RecordBatch]es as they arrive from the right  of the join.
+/// [`Stream`] for [`HashJoinExec`] that does the actual join.
+///
+/// This stream:
+///
+/// 1. Reads the entire left input and constructs a hash table
```
Comment (comphead) [inline datafusion/physical-plan/src/joins/hash_join.rs:754]:
```suggestion
/// 1. Reads the entire left input(build) and constructs a hash table
```

File: datafusion/physical-plan/src/joins/hash_join.rs [hunk 3]
```diff
@@ -747,27 +747,38 @@ where
...
+///
+/// 2. Streams [RecordBatch]es as they arrive from the right input and joins
```
Comment (comphead) [inline datafusion/physical-plan/src/joins/hash_join.rs:756]:
```suggestion
/// 2. Streams [RecordBatch]es as they arrive from the right input(probe) and joins
```

**Review by comphead — APPROVED** [review]
LGTM thanks @alamb

**Review by Dandandan — APPROVED** [review]

**Pull request merged**